File: backend/tests/factories.py
"""Deterministic object factories shared by the Phase 3 test modules.

Everything here is plain construction — no randomness beyond explicit UUID
defaults, no clock reads — so tests stay reproducible.
"""

import itertools
import uuid
from collections.abc import Callable, Sequence

from pydantic import JsonValue

from app.models.alert import Alert
from app.models.candidate_alert import CandidateAlert
from app.models.enums import Category, Severity, SourceType
from app.models.packet_event import PacketEvent

# A syntactically valid placeholder dedup key (40 lowercase hex chars).
PLACEHOLDER_DEDUP_KEY = "0" * 40

# Shared sensor token for API tests (>= 16 chars per config validation).
TEST_SENSOR_TOKEN = "unit-test-sensor-token"


def auth_headers() -> dict[str, str]:
    """The ingest authentication header for the test sensor token."""
    return {"X-Sensor-Token": TEST_SENSOR_TOKEN}


def ingest_payload(events: Sequence[PacketEvent]) -> dict[str, object]:
    """A JSON-ready ingest request body for ``events``."""
    return {"events": [event.model_dump(mode="json") for event in events]}


def sequential_id_factory(start: int = 0) -> Callable[[], str]:
    """Return an ``id_factory`` producing deterministic, distinct UUIDv4 strings."""
    counter = itertools.count(start)

    def next_id() -> str:
        return str(uuid.UUID(int=next(counter), version=4))

    return next_id


def make_candidate(
    *,
    detector_id: str = "portscan",
    detector_version: str = "1.0",
    category: Category | None = None,
    severity: Severity = "medium",
    confidence: float = 0.6,
    src_ip: str | None = "10.0.0.50",
    dst_ip: str = "10.0.0.10",
    source_type: SourceType = "synthetic",
    evidence: dict[str, JsonValue] | None = None,
    threshold_snapshot: dict[str, JsonValue] | None = None,
    window_start: float = 1000.0,
    window_end: float = 1002.0,
) -> CandidateAlert:
    """Build a valid :class:`CandidateAlert` with overridable fields."""
    if category is None:
        category = "reconnaissance" if detector_id == "portscan" else "dos"
    return CandidateAlert(
        detector_id=detector_id,
        detector_version=detector_version,
        category=category,
        severity=severity,
        confidence=confidence,
        src_ip=src_ip,
        dst_ip=dst_ip,
        source_type=source_type,
        evidence=evidence if evidence is not None else {"distinct_port_count": 15},
        threshold_snapshot=(
            threshold_snapshot if threshold_snapshot is not None else {"PORTSCAN_MIN_PORTS": 15}
        ),
        window_start=window_start,
        window_end=window_end,
    )


def make_alert(
    *,
    alert_id: str | None = None,
    created_at: float = 1000.0,
    detector_id: str = "portscan",
    detector_version: str = "1.0",
    category: Category = "reconnaissance",
    severity: Severity = "medium",
    confidence: float = 0.6,
    src_ip: str | None = "10.0.0.50",
    dst_ip: str = "10.0.0.10",
    window_start: float = 1000.0,
    window_end: float = 1002.0,
    evidence: dict[str, JsonValue] | None = None,
    threshold_snapshot: dict[str, JsonValue] | None = None,
    dedup_key: str = PLACEHOLDER_DEDUP_KEY,
    source_type: SourceType = "synthetic",
    occurrence_count: int = 1,
    last_seen: float | None = None,
) -> Alert:
    """Build a valid persisted-shape :class:`Alert` with overridable fields."""
    return Alert(
        alert_id=alert_id if alert_id is not None else str(uuid.uuid4()),
        created_at=created_at,
        detector_id=detector_id,
        detector_version=detector_version,
        category=category,
        severity=severity,
        confidence=confidence,
        src_ip=src_ip,
        dst_ip=dst_ip,
        window_start=window_start,
        window_end=window_end,
        evidence=evidence if evidence is not None else {"distinct_port_count": 15},
        threshold_snapshot=(
            threshold_snapshot if threshold_snapshot is not None else {"PORTSCAN_MIN_PORTS": 15}
        ),
        dedup_key=dedup_key,
        source_type=source_type,
        occurrence_count=occurrence_count,
        last_seen=last_seen if last_seen is not None else created_at,
    )

File: frontend/src/test/factories.ts
/** Shared test factories for building valid domain objects. */
import type { Alert } from '../types/alert.ts'
import type { StatsResponse } from '../types/stats.ts'
import type { AlertEnvelope, AlertEnvelopeType } from '../types/ws.ts'

let seq = 0

/** A deterministic UUIDv4-shaped id (unique per call unless overridden). */
export function makeAlertId(n = ++seq): string {
  const hex = n.toString(16).padStart(12, '0')
  return `11111111-1111-4111-8111-${hex}`
}

export function makeAlert(overrides: Partial<Alert> = {}): Alert {
  return {
    alert_id: makeAlertId(),
    created_at: 1000,
    detector_id: 'portscan',
    detector_version: '1.0',
    category: 'reconnaissance',
    severity: 'medium',
    confidence: 0.7,
    src_ip: '10.0.0.5',
    dst_ip: '10.0.0.9',
    window_start: 1000,
    window_end: 1010,
    evidence: { distinct_port_count: 15, sampled_ports: [22, 80, 443], duration_s: 10 },
    threshold_snapshot: { PORTSCAN_MIN_PORTS: 15, PORTSCAN_WINDOW_S: 10 },
    dedup_key: 'dedup-portscan',
    source_type: 'synthetic',
    occurrence_count: 1,
    last_seen: 1010,
    ai_explanation: null,
    ai_status: 'none',
    ...overrides,
  }
}

export function makeEnvelope(
  type: AlertEnvelopeType,
  overrides: Partial<Alert> = {},
): AlertEnvelope {
  return { type, alert: makeAlert(overrides) }
}

export function makeStats(overrides: Partial<StatsResponse> = {}): StatsResponse {
  return {
    totals: {
      alert_count: 1,
      alert_occurrence_total: 2,
      event_count: 100,
      byte_count: 6400,
    },
    alerts_by_severity: { low: 0, medium: 1, high: 0, critical: 0 },
    alerts_by_detector: { portscan: 1, synflood: 0 },
    alerts_by_source_type: { synthetic: 1, replay: 0, live: 0 },
    protocol_distribution: [{ protocol: 'TCP', packet_count: 100, byte_count: 6400 }],
    traffic_timeline: [
      { bucket_ts: 1000, protocol: 'TCP', source_type: 'synthetic', packet_count: 5, byte_count: 320 },
    ],
    ...overrides,
  }
}
